### DISCUSSION TITLE
Combine several pipelines with shared vectors

### DISCUSSION THREAD

**Discussion by Pandalei97** [explosion/spaCy#8067] [Help: Coding & Implementations] [answered]
Hi ! I trained a ner model and a textcat model separatly because they have different train/dev data. But they are all trained from the same vectors.

Now I want to add the two components to into a blank model which is initialized with the same vectors and each component will keep their own tok2vec.

Currently I have the 'vocab' directory in ner_model, textcat_model and blank_model. When I import the two ner and textcat models into blank model, are we actually using only the vocab of the blank model ? If true, is it safe to remove the vocab directory in each component to save storage space ? Unless I'm mistaken, I think the only difference among these 3 vocab directories is the strings.json file.

Thanks in advance for your response !

**Comment by adrianeboyd** [comment, marked as answer]
Yes, you have to pay attention to the vectors when sourcing components between pipelines. There is currently a logger warning if the vectors don't match, but there's nothing that prevents you from sourcing between pipelines with different vectors. When you use `.add_pipe(source=)` all the strings from the source model get added, too, since they may be required.

If `ner` and `textcat` have separate `tok2vec` components that aren't shared with any other components, it's easiest to move them into the component itself so you don't have deal with interactions/renaming.

You can use `replace_listeners` and `add_pipe` to do this programmatically:

```python
ner_model.replace_listeners("tok2vec", "ner", ["model.tok2vec"])
blank_model.add_pipe("ner", source=ner_model)
textcat_model.replace_listeners("tok2vec", "textcat", ["model.tok2vec"])
blank_model.add_pipe("textcat", source=textcat_model)
blank_model.to_disk("/path/to/model")
```

As as alternative, you can also assemble a model from a config with the same basic settings/options. Excerpts for the relevant config blocks:

```ini
[components.ner]
source = "ner_model"
replace_listeners = ["model.tok2vec"]
component = "ner"

[components.textcat]
source = "textcat_model"
replace_listeners = ["model.tok2vec"]
component = "textcat"

[initialize.before_init]
@callbacks = "spacy.copy_from_base_model.v1"
tokenizer = "blank_model"
vocab = "blank_model"
```

```bash
spacy assemble config.cfg -o /path/to/model
```
